src/Core/IdentityServer/ResourceOwnerPasswordValidator.cs

```diff
@@ -60,25 +59,36 @@ public async Task ValidateAsync(ResourceOwnerPasswordValidationContext context)
             //    return;
             //}
 
-            if (_captchaValidationService.ServiceEnabled && _currentContext.IsBot)
+            string bypassToken = null;
+            if (_captchaValidationService.ServiceEnabled && (_currentContext.IsBot || _captchaValidationService.RequireCaptcha))
             {
-                var captchaResponse = context.Request.Raw["CaptchaResponse"]?.ToString();
+                var user = await _userManager.FindByEmailAsync(context.UserName.ToLowerInvariant());
+                var captchaResponse = context.Request.Raw["captchaResponse"]?.ToString();
+
                 if (string.IsNullOrWhiteSpace(captchaResponse))
                 {
-                    context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, "Captcha required.");
+                    context.Result = new GrantValidationResult(TokenRequestErrors.InvalidGrant, "Captcha required.",
+                        new Dictionary<string, object> {
+                            { "HCaptcha_SiteKey", _captchaValidationService.SiteKey },
+                        });
                     return;
                 }
 
-                var captchaValid = await _captchaValidationService.ValidateCaptchaResponseAsync(captchaResponse,
-                    _currentContext.IpAddress);
+                var captchaValid = _captchaValidationService.ValidateCaptchaBypassToken(captchaResponse, user) ||
+                    await _captchaValidationService.ValidateCaptchaResponseAsync(captchaResponse, _currentContext.IpAddress);
                 if (!captchaValid)
                 {
-                    await BuildErrorResultAsync("Captcha is invalid.", false, context, null);
+                    await BuildErrorResultAsync("Captcha is invalid. Please refresh and try again", false, context, null);
                     return;
                 }
+                bypassToken = _captchaValidationService.GenerateCaptchaBypassToken(user);
             }
 
             await ValidateAsync(context, context.Request);
+            if (context.Result.CustomResponse != null && bypassToken != null)
+            {
+                context.Result.CustomResponse["HCaptcha_BypassKey"] = bypassToken;
```

Comment:
I'd call this CaptchaBypassToken. You call it token everywhere else and it shouldn't be HCpatcha-specific.